lib/wasix/src/syscalls/wasix/context_create.rs

```diff
@@ -0,0 +1,96 @@
+use crate::{WasiEnv, WasiError};
+use tracing::instrument;
+use wasmer::{
+    AsStoreRef, Function, FunctionEnvMut, MemorySize, RuntimeError, StoreMut, TypedFunction, Value,
+    WasmPtr,
+};
+use wasmer_wasix_types::wasi::Errno;
+
+/// Return the function corresponding to the given entrypoint index if it exists and has the signature `() -> ()`
+pub fn lookup_typechecked_entrypoint(
+    data: &WasiEnv,
+    mut store: &mut StoreMut<'_>,
+    entrypoint_id: u32,
+) -> Result<TypedFunction<(), ()>, Errno> {
+    let entrypoint = match data
+        .inner()
+        .indirect_function_table_lookup(&mut store, entrypoint_id)
+    {
+        Ok(func) => func,
+        Err(e) => {
+            tracing::trace!(
+                "Failed to lookup entrypoint function {}: {:?}",
+                entrypoint_id,
+                e
+            );
+            return Err(Errno::Inval);
+        }
+    };
+
+    let entrypoint_type = entrypoint.ty(&store);
+    let Ok(entrypoint) = entrypoint.typed::<(), ()>(&store) else {
+        tracing::trace!(
+            "Entrypoint function {entrypoint_id} has invalid signature: expected () -> (), got {:?} -> {:?}",
+            entrypoint_type.params(),
+            entrypoint_type.results()
+        );
+        return Err(Errno::Inval);
+    };
+
+    Ok(entrypoint)
+}
+
+/// Create a new context.
+///
+/// Creates a new context in the suspended state. On its first resumption,
+/// `entrypoint` is invoked within that context.
+///
+/// Refer to the wasix-libc [`wasix/context.h`] header for authoritative
+/// documentation.
+///
+/// [`wasix/context.h`]: https://github.com/wasix-org/wasix-libc/blob/main/libc-bottom-half/headers/public/wasix/context.h
+#[instrument(level = "trace", skip(ctx), ret)]
+pub fn context_create<M: MemorySize>(
+    mut ctx: FunctionEnvMut<'_, WasiEnv>,
+    new_context_ptr: WasmPtr<u64, M>,
+    entrypoint: u32,
+) -> Result<Errno, WasiError> {
+    WasiEnv::do_pending_operations(&mut ctx)?;
+
+    // Verify that we are in an async context
+    // We need to do this first, before we borrow the store mutably
+    let Some(async_store) = ctx.as_store_async() else {
+        tracing::warn!(
```

Comment:
This message mentions engine support, but we don't even expose this function to modules if the engine doesn't support async. This can happen if we launch the module synchronously which, as stated elsewhere, I don't think we should ever do after this change.